sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -364,83 +366,263 @@ struct codec_dai dai_lists[] = {
+	dai_links[id].platforms = platform_component;
+	dai_links[id].num_platforms = ARRAY_SIZE(platform_component);
+	dai_links[id].nonatomic = true;
+	dai_links[id].no_pcm = 1;
+	dai_links[id].cpus = &cpus[id];
+	dai_links[id].num_cpus = 1;
+	dai_links[id].cpus->dai_name = cpu_name;
+	dai_links[id].codecs = codecs;
+	dai_links[id].num_codecs = codecs_num;
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	unsigned int part_id, unique_id, mfg_id, link_id, class_id;
+	struct snd_soc_dai_link_component *codec;
+	int i, j, dai_num = 1;
+
+	codec = devm_kzalloc(dev, sizeof(struct snd_soc_dai_link_component) *
+			     link->num_adr, GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	/*
+	 * Spec definition
+	 *   Register		Bit	Contents
+	 *   DevId_0 [7:4]	47:44	sdw_version
+	 *   DevId_0 [3:0]	43:40	unique_id
+	 *   DevId_1		39:32	mfg_id [15:8]
+	 *   DevId_2		31:24	mfg_id [7:0]
+	 *   DevId_3		23:16	part_id [15:8]
+	 *   DevId_4		15:08	part_id [7:0]
+	 *   DevId_5		07:00	class_id
+	 */
+	link_id = (link->adr[0] >> 48) & GENMASK(15, 0);
+	mfg_id = (link->adr[0] >> 24) & GENMASK(15, 0);
+	part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+	class_id = link->adr[0] & GENMASK(7, 0);
+
+	if (link->num_adr == 1) {
+		codec[0].name = devm_kasprintf(dev, GFP_KERNEL,
+					       "sdw:%x:%x:%x:%x", link_id,
+					       mfg_id, part_id, class_id);
+
+		if (!codec[0].name)
+			return -ENOMEM;
+	} else {
+		for (i = 0; i < link->num_adr; i++) {
+			unique_id = (link->adr[i] >> 40) & GENMASK(3, 0);
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
+
+			if (!codec[i].name)
+				return -ENOMEM;
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(dai_lists); i++)
+		if (part_id == dai_lists[i].id)
+			break;
+
+	if (i == ARRAY_SIZE(dai_lists))
+		return -EINVAL;
+
+	for (j = 0; j < link->num_adr; j++)
+		codec[j].dai_name = dai_lists[i].name;
+
+	/* playback & capture are supported in different dai */
+	if (part_id == 0x711)
+		dai_num++;
+
+	j = 2;
+	for (i = id; i < dai_num + id; i++) {
+		char *name, *cpu_name;
+
+		if (i == id && part_id != 0x715) {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "SDW%d-Playback", link_id);
+			dai_links[i].dpcm_playback = 1;
+		} else {
+			name = devm_kasprintf(dev, GFP_KERNEL,
+					      "SDW%d-Capture", link_id);
+			dai_links[i].dpcm_capture = 1;
+		}
+
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j++);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, i, name, cpus, cpu_name,
+			      codec, link->num_adr);
+	}
+
+	if (part_id == 0x711)
+		dai_links[id].init = headset_init;
+
+	if ((link_id == 0x2 || link->num_adr > 1) && part_id == 0x1308)
+		dai_links[id].init = second_spk_init;
```

Comment:
The above two checks / assignments seem very specific to me... I'm definitely missing something.